balancer/rls/internal/client_test.go

```diff
@@ -0,0 +1,222 @@
+	server, cc, cleanup := setup(t)
+	defer cleanup()
+
+	// We setup the fake server to return an error.
+	server.ResponseChan <- fakeserver.Response{Err: errors.New("rls failure")}
+
+	rlsClient := newRLSClient(cc, defaultDialTarget, defaultRPCTimeout)
+
+	errCh := make(chan error)
+	cancel := rlsClient.lookup("", nil, func(target, headerData string, err error) {
+		if err == nil {
+			errCh <- errors.New("rlsClient.lookup() succeeded, should have failed")
+			return
+		}
+		if target != "" || headerData != "" {
+			errCh <- fmt.Errorf("rlsClient.lookup() = (%s, %s), should be empty strings", target, headerData)
+			return
+		}
+		errCh <- nil
+	})
+	defer cancel()
+
+	timer := time.NewTimer(defaultTestTimeout)
+	select {
+	case <-timer.C:
+		t.Fatal("Timeout when expecting a routeLookup callback")
+	case err := <-errCh:
+		timer.Stop()
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
+// TestLookupCanceled tests the case where the lookup is cancelled, and
+// verifies that the provided callback is not invoked.
+func TestLookupCanceled(t *testing.T) {
+	_, cc, cleanup := setup(t)
+	defer cleanup()
+
+	rlsClient := newRLSClient(cc, defaultDialTarget, defaultRPCTimeout)
+
+	errCh := make(chan error)
+	cancel := rlsClient.lookup("", nil, func(_, _ string, err error) {
+		errCh <- errors.New("lookup callback invoked after lookup was cancelled")
+	})
+	cancel()
+
+	timer := time.NewTimer(defaultTestTimeout)
+	select {
+	case err := <-errCh:
+		timer.Stop()
+		t.Fatal(err)
+	case <-timer.C:
+	}
+}
+
+// TestLookupDeadlineExceeded tests the case where the RPC deadline associated
+// with the lookup expires.
+func TestLookupDeadlineExceeded(t *testing.T) {
+	_, cc, cleanup := setup(t)
+	defer cleanup()
+
+	// Give the Lookup RPC a small deadline, but don't setup the fake server to
+	// return anything. So the Lookup call will block and eventuall expire.
+	rlsClient := newRLSClient(cc, defaultDialTarget, 100*time.Millisecond)
+
+	errCh := make(chan error)
+	cancel := rlsClient.lookup("", nil, func(target, headerData string, err error) {
+		if grpc.Code(err) != codes.DeadlineExceeded {
+			errCh <- fmt.Errorf("rlsClient.lookup() returned error: %v, want %v", err, codes.DeadlineExceeded)
+			return
+		}
+		errCh <- nil
+	})
+	defer cancel()
+
+	timer := time.NewTimer(defaultTestTimeout)
+	select {
+	case <-timer.C:
+		t.Fatal("Timeout when expecting a routeLookup callback")
+	case err := <-errCh:
+		timer.Stop()
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
+// TestLookupSuccess verifies the successful Lookup API case.
+func TestLookupSuccess(t *testing.T) {
+	server, cc, cleanup := setup(t)
+	defer cleanup()
+
+	const (
+		defaultTestTimeout = 1 * time.Second
+		rlsReqPath         = "/service/method"
+		rlsRespTarget      = "us_east_1.firestore.googleapis.com"
+		rlsHeaderData      = "headerData"
+	)
+
+	// We setup the fake server to return this response when it receives a
+	// request.
+	server.ResponseChan <- fakeserver.Response{
```

Comment:
Move this after checking `<-server.RequestChan`?

Reply:
Done.